File: reference-implementation/python/examples/mcp_agent_demo.py
# -*- coding: utf-8 -*-
"""
A2CN MCP Agent Demo — SaaS Renewal Negotiation via MCP Tools

Demonstrates how an LLM agent uses the A2CN MCP server tools to negotiate
commercial terms autonomously.  The buyer (TechCorp) interacts exclusively
through the six MCP tools exposed by mcp_server.py.  The seller (Acme) is
simulated in-process using MockLLM, mirroring the pattern in saas_renewal.py.

Architecture (per spec Section 13.9):
    Buyer MockLLM
        → decides: what to offer, whether to accept or reject
    MCP tool layer (mcp_server.py)
        → handles: session state, DID auth, message signing, sequence tracking
    A2CN FastAPI server (seller / responder side)
        → processes all messages, enforces protocol rules, generates records

Expected output:
    ✓ Discovered Acme: A2CN v0.2, conformance 2, deal types: ['saas_renewal']
    ✓ Session initiated — id: xxxxxxxx...
    ✓ Exchange 1: Buyer $95,000 → Seller $109,000
      [Buyer LLM] Counter: $99,000 / net-30
    ✓ Exchange 2: Buyer $99,000 → Seller $106,000
      [Buyer LLM] Accepting $106,000 / net-30
    ✓ Exchange 2: Buyer accepts — deal at $106,000 / net-30
    ✓ Transaction record_hash: ...
    ✓ A2CN MCP agent demo complete

Run with:
    cd reference-implementation/python
    python examples/mcp_agent_demo.py
"""

from __future__ import annotations

import sys
import io

if sys.stdout.encoding and sys.stdout.encoding.lower() != "utf-8":
    sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding="utf-8", errors="replace")

import os
import asyncio
import threading
import time
from pathlib import Path

# ---------------------------------------------------------------------------
# Path setup — must happen before any a2cn / mcp_server imports
# ---------------------------------------------------------------------------

_REPO_PY = Path(__file__).parent.parent          # reference-implementation/python
_EXAMPLES = Path(__file__).parent                # reference-implementation/python/examples

if str(_REPO_PY) not in sys.path:
    sys.path.insert(0, str(_REPO_PY))
if str(_EXAMPLES) not in sys.path:
    sys.path.insert(0, str(_EXAMPLES))

# ---------------------------------------------------------------------------
# Agent identity — configure via environment variables before mcp_server loads.
# mcp_server reads these at import time to build the ephemeral DID document.
# ---------------------------------------------------------------------------

os.environ.setdefault("A2CN_AGENT_DID", "did:web:mcp-buyer.demo")
os.environ.setdefault("A2CN_AGENT_ID", "mcp-buyer-001")
os.environ.setdefault("A2CN_AGENT_ORG", "TechCorp Inc")

import httpx
import uvicorn

import mcp_server as mcp_srv
from llm_agent import (
    MockLLM,
    NegotiationSkill,
    get_validated_decision,
    build_terms_from_decision,
    _build_seller_counteroffer,
)

# ---------------------------------------------------------------------------
# Demo constants
# ---------------------------------------------------------------------------

SELLER_DID = "did:web:acme-corp.com"
BUYER_DID = mcp_srv._agent_did          # resolved from env var above
SELLER_PORT = 8002
SELLER_ENDPOINT = f"http://localhost:{SELLER_PORT}"


# ---------------------------------------------------------------------------
# Patch DID-to-URL resolution for local demo
#
# In production:  did:web:acme-corp.com  →  https://acme-corp.com
# In this demo:   did:web:acme-corp.com  →  http://localhost:8002
#
# This lets the MCP tools talk to a local server without modifying mcp_server.py.
# ---------------------------------------------------------------------------

_orig_did_to_base_url = mcp_srv._did_to_base_url


def _demo_did_to_base_url(did: str) -> str:
    if did == SELLER_DID:
        return SELLER_ENDPOINT
    return _orig_did_to_base_url(did)


mcp_srv._did_to_base_url = _demo_did_to_base_url


# ---------------------------------------------------------------------------
# Server helpers (same pattern as saas_renewal.py)
# ---------------------------------------------------------------------------

def _start_server(app, port: int) -> None:
    config = uvicorn.Config(app, host="127.0.0.1", port=port, log_level="error")
    uvicorn.Server(config).run()


async def _wait_for_server(url: str, timeout: float = 10.0) -> None:
    async with httpx.AsyncClient() as http:
        deadline = time.time() + timeout
        while time.time() < deadline:
            try:
                await http.get(url)
                return
            except Exception:
                await asyncio.sleep(0.1)
    raise RuntimeError(f"Server at {url} did not start within {timeout}s")


# ---------------------------------------------------------------------------
# Main demo
# ---------------------------------------------------------------------------

async def main() -> None:
    from a2cn.crypto import generate_keypair
    import a2cn.server as server_module
    from a2cn.server import app, configure_responder, manager, register_did_document

    # -----------------------------------------------------------------------
    # 1. Configure Acme (seller) as the A2CN responder
    # -----------------------------------------------------------------------
    acme_priv, _ = generate_keypair()

    configure_responder({
        "agent_info": {
            "organization_name": "Acme Corp",
            "did": SELLER_DID,
            "verification_method": f"{SELLER_DID}#key-1",
            "agent_id": "sales-agent-acme-007",
            "endpoint": SELLER_ENDPOINT,
        },
        "mandate": {
            "mandate_type": "declared",
            "agent_id": "sales-agent-acme-007",
            "principal_organization": "Acme Corp",
            "principal_did": SELLER_DID,
            "authorized_deal_types": ["saas_renewal"],
            "max_commitment_value": 20_000_000,
            "max_commitment_currency": "USD",
            "valid_from": "2026-01-01T00:00:00Z",
            "valid_until": "2026-12-31T00:00:00Z",
        },
        "deal_types": ["saas_renewal"],
        "max_rounds_by_deal_type": {"saas_renewal": 8},
        "private_key": acme_priv,
    })

    # The server's own DID is used as the JWT audience claim.
    # Setting it equal to SELLER_DID means discovery returns the same value,
    # so the MCP tool's JWT audience resolves correctly.
    server_module.SERVER_DID = SELLER_DID

    # Register the MCP buyer's ephemeral DID document with the seller server
    # so it can verify Bearer JWT signatures on incoming requests.
    register_did_document(BUYER_DID, mcp_srv.agent_did_document)

    # Tell the MCP server the JWT audience directly, skipping a second
    # discovery round-trip inside a2cn_initiate_session.
    os.environ["A2CN_COUNTERPARTY_SERVER_DID"] = SELLER_DID

    # -----------------------------------------------------------------------
    # 2. Start the Acme (seller) server in a background thread
    # -----------------------------------------------------------------------
    t = threading.Thread(target=_start_server, args=(app, SELLER_PORT), daemon=True)
    t.start()
    await _wait_for_server(f"{SELLER_ENDPOINT}/.well-known/a2cn-agent")

    print()
    print("=" * 62)
    print("  A2CN MCP Agent Demo — SaaS Renewal Negotiation")
    print("=" * 62)
    print(f"  Buyer  (MCP agent): {BUYER_DID}")
    print(f"  Seller (Acme Corp): {SELLER_DID}")
    print(f"  Transport:          HTTP → {SELLER_ENDPOINT}")
    print()

    # -----------------------------------------------------------------------
    # 3. Discover the seller via MCP tool (Tool 1: a2cn_discover)
    # -----------------------------------------------------------------------
    disc = await mcp_srv.a2cn_discover(SELLER_DID)
    if disc.get("a2cn_capable"):
        print(
            f"✓ Discovered Acme: A2CN v{disc['a2cn_version']}, "
            f"conformance {disc['conformance_level']}, "
            f"deal types: {disc['deal_types']}"
        )
    else:
        # Graceful degradation — demo continues even if discovery HTTP fails
        print(f"  [discovery skipped — demo mode]: {disc.get('message', '')[:70]}")

    # -----------------------------------------------------------------------
    # 4. Initiate session and send opening offer (Tool 2: a2cn_initiate_session)
    # -----------------------------------------------------------------------
    buyer_opening_cents = 9_500_000   # $95,000
    buyer_opening_net   = 30

    print(f"\n[Buyer LLM] Opening offer: ${buyer_opening_cents // 100:,} / net-{buyer_opening_net}")

    init_result = await mcp_srv.a2cn_initiate_session(
        counterparty_did=SELLER_DID,
        deal_type="saas_renewal",
        my_did=BUYER_DID,
        initial_offer_total_value_cents=buyer_opening_cents,
        currency="USD",
        max_rounds=6,
        payment_terms_net_days=buyer_opening_net,
        subject="Acme Analytics Platform — FY2027 renewal",
    )

    if "error" in init_result:
        print(f"✗ Session init failed: {init_result}")
        return

    session_id = init_result["session_id"]
    print(f"✓ Session initiated — id: {session_id[:16]}...")

    # -----------------------------------------------------------------------
    # 5. Negotiation loop
    #
    # Buyer decisions come from MockLLM via MCP tools.
    # Seller decisions come from MockLLM processed directly via the server's
    # session manager (simulating the seller's in-process agent).
    # -----------------------------------------------------------------------
    buyer_skill = NegotiationSkill(
        role="buyer",
        deal_type="saas_renewal",
        floor_value_cents=10_600_000,   # max willing to pay: $106,000
        target_value_cents=buyer_opening_cents,
        max_net_days=45,
        min_net_days=0,
        walk_away_rounds=3,
        rationale_template="aggressive buyer pushing for lowest total cost",
    )
    seller_skill = NegotiationSkill(
        role="seller",
        deal_type="saas_renewal",
        floor_value_cents=10_500_000,   # min willing to accept: $105,000
        target_value_cents=11_500_000,
        max_net_days=60,
        min_net_days=30,
        walk_away_rounds=3,
        rationale_template="seller protecting margin while staying competitive",
    )

    buyer_llm  = MockLLM()
    seller_llm = MockLLM()

    buyer_history:  list[dict] = [{"value": buyer_opening_cents, "net_days": buyer_opening_net}]
    seller_history: list[dict] = []

    session_obj  = manager.get_session(session_id)
    client_state = mcp_srv._sessions[session_id]["client"]._sessions[session_id]

    exchange = 1

    while True:
        # -------------------------------------------------------------------
        # Seller's turn (simulated in-process, not via HTTP)
        # -------------------------------------------------------------------
        buyer_last_offer   = client_state["latest_offer"]
        seller_offer_terms = buyer_last_offer.get("terms", {})

        seller_decision = get_validated_decision(
            seller_llm, seller_skill,
            offer_terms=seller_offer_terms,
            my_history=seller_history,
        )

        if seller_decision is None or seller_decision["action"] == "withdraw":
            print(f"\n✓ Exchange {exchange}: Seller withdrew — no deal.")
            break

        if seller_decision["action"] == "reject":
            print(f"\n✓ Exchange {exchange}: Seller rejected — no deal.")
            break

        if seller_decision["action"] == "accept":
            # Seller accepts buyer's current offer (unusual path — buyer opened aggressively)
            val = seller_offer_terms.get("total_value", 0)
            print(f"\n✓ Exchange {exchange}: Seller accepts buyer's ${val // 100:,} — deal!")
            mcp_srv._sessions[session_id]["status"] = "COMPLETED"
            break

        # Seller sends counteroffer — build and inject into both the server
        # session manager and the MCP client state
        seller_terms = build_terms_from_decision(
            seller_decision, seller_skill, seller_offer_terms
        )

        # next_seq: client_state["sequence_number"] was last incremented by
        # the buyer's most recent send_offer; seller's reply is always +1
        next_seq = client_state["sequence_number"] + 1

        seller_co = _build_seller_counteroffer(
            session_id=session_id,
            round_number=next_seq,      # round == seq (saas_renewal.py convention)
            sequence_number=next_seq,
            terms=seller_terms,
            in_reply_to=buyer_last_offer["message_id"],
            seller_did=SELLER_DID,
            seller_priv=acme_priv,
        )
        # Update server-side session manager (seller "received" the message)
        manager.process_message(session_obj, seller_co)
        # Update MCP client state so a2cn_accept / a2cn_send_offer see the offer
        mcp_srv.inject_counterparty_offer(session_id, seller_co)

        seller_val = seller_terms["total_value"]
        seller_net = seller_terms["payment_terms"]["net_days"]
        buyer_val  = seller_offer_terms.get("total_value", 0)

        print(
            f"\n✓ Exchange {exchange}: "
            f"Buyer ${buyer_val // 100:,} → "
            f"Seller ${seller_val // 100:,} / net-{seller_net}"
        )
        seller_history.append({"value": seller_val, "net_days": seller_net})

        # -------------------------------------------------------------------
        # Buyer's turn (via MCP tools + MockLLM)
        # -------------------------------------------------------------------
        buyer_decision = get_validated_decision(
            buyer_llm, buyer_skill,
            offer_terms=seller_terms,
            my_history=buyer_history,
        )

        if buyer_decision is None or buyer_decision["action"] == "withdraw":
            print("  [Buyer LLM] Withdrawing — no deal.")
            await mcp_srv.a2cn_reject(session_id)
            break

        if buyer_decision["action"] == "reject":
            print("  [Buyer LLM] Rejecting — no deal.")
            await mcp_srv.a2cn_reject(session_id)
            break

        if buyer_decision["action"] == "accept":
            print(f"  [Buyer LLM] Accepting ${seller_val // 100:,} / net-{seller_net}")

            # Tool 4: a2cn_accept
            accept_result = await mcp_srv.a2cn_accept(session_id)
            if "error" in accept_result:
                print(f"✗ Accept failed: {accept_result}")
            else:
                rec_hash = accept_result.get("record_hash", "")
                print(
                    f"\n✓ Exchange {exchange}: Buyer accepts — "
                    f"deal at ${seller_val // 100:,} / net-{seller_net}"
                )
                print(f"✓ Transaction record_hash: {rec_hash[:32]}...")
            break

        # Buyer sends counteroffer via MCP tool (Tool 3: a2cn_send_offer)
        new_val = buyer_decision["total_value_cents"]
        new_net = buyer_decision["net_days"]
        print(f"  [Buyer LLM] Counter: ${new_val // 100:,} / net-{new_net}")

        # Tool 3: a2cn_send_offer
        co_result = await mcp_srv.a2cn_send_offer(
            session_id=session_id,
            total_value_cents=new_val,
            payment_terms_net_days=new_net,
        )
        if "error" in co_result:
            print(f"✗ Send offer failed: {co_result}")
            break

        buyer_history.append({"value": new_val, "net_days": new_net})
        exchange += 1

    # -----------------------------------------------------------------------
    # 6. Final session status (Tool 6: a2cn_get_session_status)
    # -----------------------------------------------------------------------
    final = await mcp_srv.a2cn_get_session_status(session_id)
    print()
    print(f"✓ Final session state: {final['status']}")
    if final.get("transaction_record"):
        tr = final["transaction_record"]
        print(f"✓ Record ID:   {tr.get('record_id', '')}")
        print(f"  Record hash: {tr.get('record_hash', '')[:32]}...")
    print(f"✓ A2CN MCP agent demo complete")


if __name__ == "__main__":
    asyncio.run(main())

File: a2cn_ts/examples/mcp_agent_demo.ts
/**
 * A2CN MCP Agent Demo — SaaS Renewal Negotiation via MCP Tools
 *
 * Demonstrates how an LLM agent uses the A2CN MCP server tools to negotiate
 * commercial terms autonomously.  The buyer (TechCorp) interacts exclusively
 * through the six MCP tools exposed by mcp_server.  The seller (Acme) is
 * simulated in-process using MockLLM, mirroring the pattern in saas_renewal.
 *
 * Architecture (per spec Section 13.9):
 *     Buyer MockLLM
 *         → decides: what to offer, whether to accept or reject
 *     MCP tool layer (mcp_server)
 *         → handles: session state, DID auth, message signing, sequence tracking
 *     A2CN Fastify server (seller / responder side)
 *         → processes all messages, enforces protocol rules, generates records
 *
 * Expected output:
 *     ✓ Discovered Acme: A2CN v0.2, conformance 2, deal types: ['saas_renewal']
 *     ✓ Session initiated — id: xxxxxxxx...
 *     ✓ Exchange 1: Buyer $95,000 → Seller $109,000
 *       [Buyer LLM] Counter: $99,000 / net-30
 *     ✓ Exchange 2: Buyer $99,000 → Seller $106,000
 *       [Buyer LLM] Accepting $106,000 / net-30
 *     ✓ Exchange 2: Buyer accepts — deal at $106,000 / net-30
 *     ✓ Transaction record_hash: ...
 *     ✓ A2CN MCP agent demo complete
 *
 * Run with:
 *     cd a2cn_ts
 *     npx tsx examples/mcp_agent_demo.ts
 */

// ---------------------------------------------------------------------------
// Agent identity — configure via environment variables before mcp_server loads.
// The MCP context reads these at creation time to build the ephemeral DID document.
// ---------------------------------------------------------------------------

process.env.A2CN_AGENT_DID = process.env.A2CN_AGENT_DID ?? "did:web:mcp-buyer.demo";
process.env.A2CN_AGENT_ID = process.env.A2CN_AGENT_ID ?? "mcp-buyer-001";
process.env.A2CN_AGENT_ORG = process.env.A2CN_AGENT_ORG ?? "TechCorp Inc";

import { generateKeypair } from "../src/a2cn/crypto.js";
import type { Dict } from "../src/a2cn/messages.js";
import { createServerContext } from "../src/a2cn/server.js";
import { createMcpContext } from "../src/mcp_server.js";
import {
  MockLLM,
  NegotiationSkill,
  getValidatedDecision,
  buildTermsFromDecision,
  buildSellerCounteroffer,
  type LlmHistoryEntry,
} from "./llm_agent.js";

// ---------------------------------------------------------------------------
// Demo constants
// ---------------------------------------------------------------------------

const SELLER_DID = "did:web:acme-corp.com";
const SELLER_PORT = 8002;
const SELLER_ENDPOINT = `http://localhost:${SELLER_PORT}`;

// ---------------------------------------------------------------------------
// Patch DID-to-URL resolution for local demo
//
// In production:  did:web:acme-corp.com  →  https://acme-corp.com
// In this demo:   did:web:acme-corp.com  →  http://localhost:8002
//
// The MCP tools resolve the seller DID to https://acme-corp.com; this fetch
// wrapper rewrites those URLs to the local server without modifying mcp_server.
// ---------------------------------------------------------------------------

const demoFetch: typeof fetch = (async (url: string | URL | Request, init?: RequestInit) => {
  const rewritten = String(url).replace("https://acme-corp.com", SELLER_ENDPOINT);
  return fetch(rewritten, init);
}) as typeof fetch;

// ---------------------------------------------------------------------------
// Main demo
// ---------------------------------------------------------------------------

async function main(): Promise<void> {
  const mcpCtx = createMcpContext({ fetchFn: demoFetch });
  const BUYER_DID = mcpCtx.agentDid; // resolved from env var above

  // -----------------------------------------------------------------------
  // 1. Configure Acme (seller) as the A2CN responder
  // -----------------------------------------------------------------------
  const serverCtx = createServerContext();
  const { privateKey: acmePriv } = generateKeypair();

  serverCtx.configureResponder({
    agent_info: {
      organization_name: "Acme Corp",
      did: SELLER_DID,
      verification_method: `${SELLER_DID}#key-1`,
      agent_id: "sales-agent-acme-007",
      endpoint: SELLER_ENDPOINT,
    },
    mandate: {
      mandate_type: "declared",
      agent_id: "sales-agent-acme-007",
      principal_organization: "Acme Corp",
      principal_did: SELLER_DID,
      authorized_deal_types: ["saas_renewal"],
      max_commitment_value: 20_000_000,
      max_commitment_currency: "USD",
      valid_from: "2026-01-01T00:00:00Z",
      valid_until: "2026-12-31T00:00:00Z",
    },
    deal_types: ["saas_renewal"],
    max_rounds_by_deal_type: { saas_renewal: 8 },
    private_key: acmePriv,
  });

  // The server's own DID is used as the JWT audience claim.
  // Setting it equal to SELLER_DID means discovery returns the same value,
  // so the MCP tool's JWT audience resolves correctly.
  serverCtx.SERVER_DID = SELLER_DID;

  // Register the MCP buyer's ephemeral DID document with the seller server
  // so it can verify Bearer JWT signatures on incoming requests.
  serverCtx.registerDidDocument(BUYER_DID, mcpCtx.agentDidDocument);

  // Tell the MCP server the JWT audience directly, skipping a second
  // discovery round-trip inside a2cn_initiate_session.
  process.env.A2CN_COUNTERPARTY_SERVER_DID = SELLER_DID;

  // -----------------------------------------------------------------------
  // 2. Start the Acme (seller) server
  // -----------------------------------------------------------------------
  await serverCtx.app.listen({ host: "127.0.0.1", port: SELLER_PORT });

  console.log();
  console.log("=".repeat(62));
  console.log("  A2CN MCP Agent Demo — SaaS Renewal Negotiation");
  console.log("=".repeat(62));
  console.log(`  Buyer  (MCP agent): ${BUYER_DID}`);
  console.log(`  Seller (Acme Corp): ${SELLER_DID}`);
  console.log(`  Transport:          HTTP → ${SELLER_ENDPOINT}`);
  console.log();

  // -----------------------------------------------------------------------
  // 3. Discover the seller via MCP tool (Tool 1: a2cn_discover)
  // -----------------------------------------------------------------------
  const disc = await mcpCtx.a2cnDiscover(SELLER_DID);
  if (disc.a2cn_capable) {
    console.log(
      `✓ Discovered Acme: A2CN v${disc.a2cn_version}, ` +
        `conformance ${disc.conformance_level}, ` +
        `deal types: ${JSON.stringify(disc.deal_types)}`,
    );
  } else {
    // Graceful degradation — demo continues even if discovery HTTP fails
    console.log(`  [discovery skipped — demo mode]: ${String(disc.message ?? "").slice(0, 70)}`);
  }

  // -----------------------------------------------------------------------
  // 4. Initiate session and send opening offer (Tool 2: a2cn_initiate_session)
  // -----------------------------------------------------------------------
  const buyerOpeningCents = 9_500_000; // $95,000
  const buyerOpeningNet = 30;

  console.log(
    `\n[Buyer LLM] Opening offer: $${Math.trunc(buyerOpeningCents / 100).toLocaleString("en-US")} / net-${buyerOpeningNet}`,
  );

  const initResult = await mcpCtx.a2cnInitiateSession({
    counterparty_did: SELLER_DID,
    deal_type: "saas_renewal",
    my_did: BUYER_DID,
    initial_offer_total_value_cents: buyerOpeningCents,
    currency: "USD",
    max_rounds: 6,
    payment_terms_net_days: buyerOpeningNet,
    subject: "Acme Analytics Platform — FY2027 renewal",
  });

  if ("error" in initResult) {
    console.log(`✗ Session init failed: ${JSON.stringify(initResult)}`);
    await serverCtx.app.close();
    return;
  }

  const sessionId = initResult.session_id as string;
  console.log(`✓ Session initiated — id: ${sessionId.slice(0, 16)}...`);

  // -----------------------------------------------------------------------
  // 5. Negotiation loop
  //
  // Buyer decisions come from MockLLM via MCP tools.
  // Seller decisions come from MockLLM processed directly via the server's
  // session manager (simulating the seller's in-process agent).
  // -----------------------------------------------------------------------
  const buyerSkill: NegotiationSkill = {
    role: "buyer",
    deal_type: "saas_renewal",
    floor_value_cents: 10_600_000, // max willing to pay: $106,000
    target_value_cents: buyerOpeningCents,
    max_net_days: 45,
    min_net_days: 0,
    walk_away_rounds: 3,
    rationale_template: "aggressive buyer pushing for lowest total cost",
  };
  const sellerSkill: NegotiationSkill = {
    role: "seller",
    deal_type: "saas_renewal",
    floor_value_cents: 10_500_000, // min willing to accept: $105,000
    target_value_cents: 11_500_000,
    max_net_days: 60,
    min_net_days: 30,
    walk_away_rounds: 3,
    rationale_template: "seller protecting margin while staying competitive",
  };

  const buyerLlm = new MockLLM();
  const sellerLlm = new MockLLM();

  const buyerHistory: LlmHistoryEntry[] = [
    { value: buyerOpeningCents, net_days: buyerOpeningNet, rationale: "", round: 1 },
  ];
  const sellerHistory: LlmHistoryEntry[] = [];

  const sessionObj = serverCtx.manager.getSession(sessionId)!;
  const clientState = mcpCtx.sessions[sessionId].client._sessions[sessionId];

  let exchange = 1;

  for (;;) {
    // -------------------------------------------------------------------
    // Seller's turn (simulated in-process, not via HTTP)
    // -------------------------------------------------------------------
    const buyerLastOffer = clientState.latest_offer as Dict;
    const sellerOfferTerms = (buyerLastOffer.terms as Dict) ?? {};

    const sellerDecision = getValidatedDecision(
      sellerLlm,
      sellerSkill,
      sellerOfferTerms,
      sellerHistory,
    );

    if (sellerDecision === null || sellerDecision.action === "withdraw") {
      console.log(`\n✓ Exchange ${exchange}: Seller withdrew — no deal.`);
      break;
    }

    if (sellerDecision.action === "reject") {
      console.log(`\n✓ Exchange ${exchange}: Seller rejected — no deal.`);
      break;
    }

    if (sellerDecision.action === "accept") {
      // Seller accepts buyer's current offer (unusual path — buyer opened aggressively)
      const val = (sellerOfferTerms.total_value as number) ?? 0;
      console.log(
        `\n✓ Exchange ${exchange}: Seller accepts buyer's $${Math.trunc(val / 100).toLocaleString("en-US")} — deal!`,
      );
      mcpCtx.sessions[sessionId].status = "COMPLETED";
      break;
    }

    // Seller sends counteroffer — build and inject into both the server
    // session manager and the MCP client state
    const sellerTerms = buildTermsFromDecision(sellerDecision, sellerSkill, sellerOfferTerms);

    // next_seq: clientState.sequence_number was last incremented by
    // the buyer's most recent sendOffer; seller's reply is always +1
    const nextSeq = clientState.sequence_number + 1;

    const sellerCo = buildSellerCounteroffer(
      sessionId,
      nextSeq, // round == seq (saas_renewal convention)
      nextSeq,
      sellerTerms,
      buyerLastOffer.message_id as string,
      SELLER_DID,
      acmePriv,
    );
    // Update server-side session manager (seller "received" the message)
    serverCtx.manager.processMessage(sessionObj, sellerCo);
    // Update MCP client state so a2cn_accept / a2cn_send_offer see the offer
    mcpCtx.injectCounterpartyOffer(sessionId, sellerCo);

    const sellerVal = sellerTerms.total_value as number;
    const sellerNet = ((sellerTerms.payment_terms as Dict) ?? {}).net_days as number;
    const buyerVal = (sellerOfferTerms.total_value as number) ?? 0;

    console.log(
      `\n✓ Exchange ${exchange}: ` +
        `Buyer $${Math.trunc(buyerVal / 100).toLocaleString("en-US")} → ` +
        `Seller $${Math.trunc(sellerVal / 100).toLocaleString("en-US")} / net-${sellerNet}`,
    );
    sellerHistory.push({ value: sellerVal, net_days: sellerNet, rationale: "", round: nextSeq });

    // -------------------------------------------------------------------
    // Buyer's turn (via MCP tools + MockLLM)
    // -------------------------------------------------------------------
    const buyerDecision = getValidatedDecision(buyerLlm, buyerSkill, sellerTerms, buyerHistory);

    if (buyerDecision === null || buyerDecision.action === "withdraw") {
      console.log("  [Buyer LLM] Withdrawing — no deal.");
      await mcpCtx.a2cnReject(sessionId);
      break;
    }

    if (buyerDecision.action === "reject") {
      console.log("  [Buyer LLM] Rejecting — no deal.");
      await mcpCtx.a2cnReject(sessionId);
      break;
    }

    if (buyerDecision.action === "accept") {
      console.log(
        `  [Buyer LLM] Accepting $${Math.trunc(sellerVal / 100).toLocaleString("en-US")} / net-${sellerNet}`,
      );

      // Tool 4: a2cn_accept
      const acceptResult = await mcpCtx.a2cnAccept(sessionId);
      if ("error" in acceptResult) {
        console.log(`✗ Accept failed: ${JSON.stringify(acceptResult)}`);
      } else {
        const recHash = (acceptResult.record_hash as string) ?? "";
        console.log(
          `\n✓ Exchange ${exchange}: Buyer accepts — ` +
            `deal at $${Math.trunc(sellerVal / 100).toLocaleString("en-US")} / net-${sellerNet}`,
        );
        console.log(`✓ Transaction record_hash: ${recHash.slice(0, 32)}...`);
      }
      break;
    }

    // Buyer sends counteroffer via MCP tool (Tool 3: a2cn_send_offer)
    const newVal = buyerDecision.total_value_cents as number;
    const newNet = buyerDecision.net_days as number;
    console.log(
      `  [Buyer LLM] Counter: $${Math.trunc(newVal / 100).toLocaleString("en-US")} / net-${newNet}`,
    );

    // Tool 3: a2cn_send_offer
    const coResult = await mcpCtx.a2cnSendOffer({
      session_id: sessionId,
      total_value_cents: newVal,
      payment_terms_net_days: newNet,
    });
    if ("error" in coResult) {
      console.log(`✗ Send offer failed: ${JSON.stringify(coResult)}`);
      break;
    }

    buyerHistory.push({ value: newVal, net_days: newNet, rationale: "", round: exchange + 1 });
    exchange += 1;
  }

  // -----------------------------------------------------------------------
  // 6. Final session status (Tool 6: a2cn_get_session_status)
  // -----------------------------------------------------------------------
  const final = await mcpCtx.a2cnGetSessionStatus(sessionId);
  console.log();
  console.log(`✓ Final session state: ${final.status}`);
  if (final.transaction_record) {
    const tr = final.transaction_record as Dict;
    console.log(`✓ Record ID:   ${tr.record_id ?? ""}`);
    console.log(`  Record hash: ${String(tr.record_hash ?? "").slice(0, 32)}...`);
  }
  console.log("✓ A2CN MCP agent demo complete");

  await serverCtx.app.close();
}

main().catch((exc) => {
  console.error(exc);
  process.exit(1);
});
